dask/array.py

```diff
@@ -240,3 +241,299 @@ def concatenate(arrays, axis=0):
+                        
+
+    #Step 1: All of our input variables should now be setup correctly,
+    #  so figure out where to start and stop based on the blocksize
+    #  of the input data
+    leftmost_block = start/blockshape[0]
+    #We subtract 1 from stop so that if the rightmost index
+    #falls on a block boundary, we won't include a slice in the next block
+    #  i.e. slice(0,20) on blockshape(20,) should give us
+    #  (x,0), slice(0,20)
+    #  NOT
+    #  ((x,0), slice(0,20)), ((x,1), slice(0,0))
+    rightmost_block = (stop-1)/blockshape[0]
+    leftmost_block_start_index = start % blockshape[0]
+    rightmost_block_stop_index = stop - (rightmost_block*blockshape[0])
+    num_blocks = rightmost_block - leftmost_block
+
+    
+        
+    dask = {}
+    #There are three main cases to account for when doing the slicing
+    #  within a dask graph
+    #1. slicing within a single block
+    #2. slicing with only adjacent blocks, i.e. blocks 1,2
+    #3. slicing with multiple contiguous blocks, i.e. block 1,2,3,4
+    if leftmost_block == rightmost_block:    
+        dask[(out_name, 0)] = (operator.getitem, (in_name, leftmost_block), slice(leftmost_block_start_index, rightmost_block_stop_index))
+    else:
+        dask[(out_name, 0)] = (operator.getitem, (in_name, leftmost_block), slice(leftmost_block_start_index, blockshape[0]))
+        dask[(out_name, num_blocks)] = (operator.getitem, (in_name, rightmost_block), slice(0, rightmost_block_stop_index))
+            
+    for block_num in range(1, num_blocks):
+        dask[(out_name, block_num)] = (operator.getitem, (in_name, leftmost_block+block_num), slice(0, blockshape[0]))
+
+    return dask
+
+
+
+def _slice_1d(dim_shape, blocksize, index):
+    """
+    Parameters
+    ----
+    dim_shape - the number of elements in this dimension
+    blocksize - the number of elements per block in this dimension
+    index - a description of the elements in this dimension that we want
+      This might be an integer, a slice(), or an Ellipsis
+
+    Returns
+    ----
+    a dictionary where the keys are the integer indexes of the blocks that should be sliced
+      and the values are the slices
+
+
+    >>> _slice_1d(100, 20, slice(0, 35))
+    {0: slice(0, 20, None), 1: slice(0, 15, None)}
+    >>> _slice_1d(100, 20, slice(10,15))
+    {0: slice(10, 15, None)}
+    >>> _slice_1d(100, 20, slice(40, 100))
+    {2: slice(0, 20, None), 3: slice(0, 20, None), 4: slice(0, 20, None)}
+    """
+    #integer division often won't tell us how many blocks
+    #  we have.
+    num_blocks = int(ceil(float(dim_shape)/blocksize))
+    
+    if index == Ellipsis or index == slice(None, None, None):
+        return {i: index for i in range(num_blocks)}
+
+    if isinstance(index, int):
+        #integer math here is ok.
+        if index >= dim_shape:
+            raise IndexError("index %s is out of bounds for shape %s" % (index, dim_shape))
+        return {index/blocksize : index % blocksize}
```

Comment:
Assuming integer division will probably fail in python 3.  Might have to use `//` or explicit call to `math.floor`